### PR TITLE
Issue/31

### PR DESCRIPTION
**Pull request by ghost**
Reworking of the communications code. Now uses packet objects instead of structs. Transmit buffers now handle storing and conversions of packets and data.

### CODE REVIEW

File: FESS-GUI/commoninterfaceselector.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ void CommonInterfaceSelector::demoButtonCancelClicked()
 
 void CommonInterfaceSelector::serialButtonSetClicked()
 {
-    SerialDevice* deviceInterface = NULL;
+    SerialDevice* deviceInterface = nullptr;
```
Comment (JCJutson) [inline FESS-GUI/commoninterfaceselector.cpp:97]: What is nullptr? I don't see where it is defined

File: FESS-GUI/conversions.h [hunk 2]
```diff
@@ -1,22 +1,31 @@
 #ifndef CONVERSIONS_H
 #define CONVERSIONS_H
 
-#include <datatypes.h>
+#define TAU	6.283185307179586476925286766559005768394
+
+#include "cstring"
```
Comment (JCJutson) [inline FESS-GUI/conversions.h]: What is cstring? I don't think we have a file named that

File: FESS-GUI/demodevice.cpp [hunk 3]
```diff
@@ -18,41 +18,46 @@ DemoDevice::~DemoDevice()
...
 //---------------------------------------------------------------------
 // Command Interpretation
 //---------------------------------------------------------------------
 
-    bool loop = true;
-
-    while(!tx.empty() && loop)
+void DemoDevice::syncTX()
+{
+    if(loop == true)
```
Comment (JCJutson) [inline FESS-GUI/demodevice.cpp:28]: Not a big deal, but note that comparing with true is redundant. You could just as easily write if(loop) and while(tx.packetsAvailable()). It's arguably more readable. My C# project over the summer would warn you if you did this

File: FESS-GUI/flypacket.cpp [hunk 4]
```diff
@@ -1,184 +1,201 @@
...
+        case IDM_SEND_ROTATIONAL_POSITION_Y:    invalidCommand = false; break;
+        default:                                invalidCommand = true; break;
+    }
+}
 
-        else
+void FlyPacket::setCommand(FlyByte generalByte)
+{
+    commandByte = generalByte;
+    acknownledgeByte = (generalByte | IDM_CMD_DIFFERENCE);
```
Comment (JCJutson) [inline FESS-GUI/flypacket.cpp]: typo: acknowledgeByte

File: FESS-GUI/flypacket.cpp [hunk 5]
```diff
@@ -1,184 +1,201 @@
...
+}
+
+FlyByte FlyPacket::getCommand()
+{
+    return commandByte;
+}
+
+FlyByte FlyPacket::readByte()
+{
+    FlyByte returnByte = 0;
```
Comment (JCJutson) [inline FESS-GUI/flypacket.cpp]: Is there a reason to initialize this? Looks like it is set to something else in every possible code path

File: FESS-GUI/flypacket.h [hunk 6]
```diff
@@ -1,8 +1,118 @@
...
+
+    FlyByte getMaxSize();
+
+   private:
+    bool invalidCommand;
+    bool readComplete;
+    bool writeComplete;
+
+    FlyByte commandByte;
+    FlyByte acknownledgeByte;
```
Comment (JCJutson) [inline FESS-GUI/flypacket.h]: There's that typo again haha

File: FESS-GUI/flyqueue.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        void clear();
+        void reset();
+        void setSize(int);
+
+        bool isEmpty();
+
+        FlyPacket pop();
+        void push(FlyPacket);
+
+        FlyPacket operator[](const unsigned int);
```
Comment (JCJutson) [inline FESS-GUI/flyqueue.h:22]: This is the first time I've actually seen someone define an operator for a class. Very cool.

File: FESS-GUI/flywheeloperation.cpp [hunk 8]
```diff
@@ -99,59 +71,59 @@ void FlywheelOperation::setInterface(CommonDeviceInterface* deviceInterface)
  * \param acceleration The acceleration to set the flywheel to.
  * \param jerk The jerk to set the flywheel to.
```
Comment (JCJutson) [inline FESS-GUI/flywheeloperation.cpp:72]: Rename the \param jerk to \param jerkValue here in the documentation, and for the other parameter names changed.

File: FESS-GUI/commoninterfaceselector.cpp (hunk 1 above)
Comment (JCJutson) [reply FESS-GUI/commoninterfaceselector.cpp:97]: Cool. I just looked this up. Didn't know that C++11 implemented nullptr

File: FESS-GUI/flywheeloperation.cpp [hunk 9]
```diff
@@ -99,59 +71,59 @@ void FlywheelOperation::setInterface(CommonDeviceInterface* deviceInterface)
...
  */
-void FlywheelOperation::setMotion(float velocity, float acceleration, float jerk)
+void FlywheelOperation::setMotion(float velocityValue, float accelerationValue, float jerkValue)
```
Comment (JCJutson) [inline FESS-GUI/flywheeloperation.cpp]: A little confusing that these params are named the same as the private variables defined in the header.

File: FESS-GUI/flypacket.h [hunk 10]
```diff
@@ -1,8 +1,118 @@
...
+#define PACKET_END              HEADER_SIZE + MAX_PAYLOAD
+
+#define DATA_BEGINNING          HEADER_SIZE
+#define DATA_END                PACKET_END - FOOTER_SIZE
+
+#define PACKET_SIZE             PACKET_END + FOOTER_SIZE
+
+// Error Types
+//------------------------------------------------
+#define NO_ERROR                0
```
Comment (JCJutson) [inline FESS-GUI/flypacket.h]: screenshot 13
I know you don't want to see this, but there is a warning when this is compiled on windows: "NO_ERROR" redefined

**Review by JCJutson — CHANGES_REQUESTED** [review]
Some changes requested. If you don't have time to address them, we can just push this through since there is nothing build-breaking

File: FESS-GUI/demodevice.cpp (hunk 3 above)
Comment (JCJutson) [reply FESS-GUI/demodevice.cpp:28]: I'm not going to hold up the PR over this though

File: FESS-GUI/conversions.h (hunk 2 above)
Comment (ghost) [reply FESS-GUI/conversions.h]: Oops... it should be &lt;cstring&gt;

**Review by JCJutson — APPROVED** [review]
Good stuff

**Pull request merged**